File: dashboard/src/auth/passwordPolicy.ts
/**
 * A client-side mirror of the server's password policy.
 *
 * This exists to tell someone what is wrong *while they are typing*, not to
 * decide whether the password is acceptable. The server re-validates every
 * request and is the only authority; this file is a convenience that saves a
 * round trip and a rejected form. If the two ever disagree, the server wins and
 * the person sees the server's message.
 *
 * Kept in step with `PasswordPolicy` in the domain. The rule weights length over
 * composition -- a 12-character floor, no digit/symbol/case requirements --
 * because composition rules reliably produce "Password1!" while length is the
 * property that actually costs an attacker something.
 */

/** Matches PasswordPolicy.MinimumLength. */
export const PASSWORD_MINIMUM_LENGTH = 12

/** Matches PasswordPolicy.MaximumLength; bounds hasher work rather than discouraging passphrases. */
export const PASSWORD_MAXIMUM_LENGTH = 256

/**
 * Returns a reason the password is unacceptable, or null when it looks fine.
 *
 * Returns the first problem rather than a list: the reader is mid-way through
 * retyping a password, and a wall of simultaneous complaints is harder to act on
 * than one instruction.
 */
export function validateNewPassword(password: string): string | null {
  if (password.trim().length === 0) {
    return 'A password is required.'
  }

  // Array.from, not .length: a string's length counts UTF-16 code units, so an
  // emoji or an astral-plane character would count as two and let a shorter
  // password through than the server will accept.
  const characters = Array.from(password).length

  if (characters < PASSWORD_MINIMUM_LENGTH) {
    return `The password must be at least ${PASSWORD_MINIMUM_LENGTH} characters.`
  }

  if (characters > PASSWORD_MAXIMUM_LENGTH) {
    return `The password must be at most ${PASSWORD_MAXIMUM_LENGTH} characters.`
  }

  if (new Set(Array.from(password)).size === 1) {
    return 'The password must not be a single repeated character.'
  }

  return null
}

/**
 * Everything wrong with the form, for enabling the submit button.
 *
 * The "same as current" check is a courtesy: the server refuses it too, because
 * re-setting the same password would rotate the security stamp and destroy every
 * session for no security gain.
 */
export function validateChangePasswordForm(input: {
  currentPassword: string
  newPassword: string
  confirmPassword: string
}): { newPassword: string | null; confirmPassword: string | null; canSubmit: boolean } {
  const newPassword = input.newPassword.length > 0 ? validateNewPassword(input.newPassword) : null

  const confirmPassword =
    input.confirmPassword.length > 0 && input.confirmPassword !== input.newPassword
      ? 'The passwords do not match.'
      : null

  const sameAsCurrent =
    input.newPassword.length > 0 && input.newPassword === input.currentPassword
      ? 'The new password must be different from the current one.'
      : null

  const canSubmit =
    input.currentPassword.length > 0 &&
    input.newPassword.length > 0 &&
    input.confirmPassword === input.newPassword &&
    validateNewPassword(input.newPassword) === null &&
    sameAsCurrent === null

  return { newPassword: newPassword ?? sameAsCurrent, confirmPassword, canSubmit }
}

File: server/Domain/Identity/PasswordPolicy.cs
namespace EndpointPlatform.Domain.Identity;

/// <summary>
/// What the platform will accept as an administrator password.
/// </summary>
/// <remarks>
/// <para>
/// A pure rule, deliberately separate from hashing and from the change-password
/// flow, so that "what counts as acceptable" can be asserted without a database,
/// an HTTP request, or a hasher.
/// </para>
/// <para>
/// <b>Length is weighted over composition.</b> The minimum is 12 characters, and
/// there is no requirement for a digit, a symbol, or mixed case. Composition
/// rules reliably produce <c>Password1!</c> and a sticky note; length is the
/// property that actually costs an attacker something. This follows current NIST
/// 800-63B guidance rather than the older complexity-rule tradition, and it
/// matches the 12-character floor the bootstrapper already enforces.
/// </para>
/// <para>
/// The upper bound exists to stop a password long enough to be a denial of
/// service against the hasher, not because long passphrases are undesirable.
/// </para>
/// </remarks>
public static class PasswordPolicy
{
    /// <summary>The same floor the bootstrap tool enforces, so the two cannot disagree.</summary>
    public const int MinimumLength = 12;

    /// <summary>
    /// Bounded so a caller cannot make the hasher do unbounded work.
    /// </summary>
    /// <remarks>
    /// Argon2/PBKDF2-class hashing is intentionally expensive; feeding it a
    /// megabyte of input turns that cost into an availability problem. 256 is far
    /// above any real passphrase.
    /// </remarks>
    public const int MaximumLength = 256;

    /// <summary>
    /// Returns a refusal reason, or null when the password is acceptable.
    /// </summary>
    /// <remarks>
    /// Returns the first failure rather than a list. The caller shows this to a
    /// person who is retyping a password, and a wall of simultaneous complaints is
    /// harder to act on than one clear instruction.
    /// </remarks>
    public static string? Validate(string? password)
    {
        if (string.IsNullOrWhiteSpace(password))
        {
            return "A password is required.";
        }

        // Length in characters, not bytes. A passphrase in a non-Latin script
        // would otherwise clear a byte-based floor on far fewer characters.
        if (password.Length < MinimumLength)
        {
            return $"The password must be at least {MinimumLength} characters.";
        }

        if (password.Length > MaximumLength)
        {
            return $"The password must be at most {MaximumLength} characters.";
        }

        // Whitespace-only was caught above; a password that is merely one
        // character repeated clears the length floor while carrying almost no
        // entropy, and is a common way to satisfy a length rule without meaning
        // to choose a password at all.
        if (password.Distinct().Count() == 1)
        {
            return "The password must not be a single repeated character.";
        }

        return null;
    }
}
